File: src/test/java/ChainOfResponsibilityTests.java
import com.dambra.paul.moduluschecker.chain.ModulusChainLink;
import com.dambra.paul.moduluschecker.ModulusCheckParams;
import com.dambra.paul.moduluschecker.chain.ModulusResult;
import org.junit.Test;

import java.util.Optional;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;

public class ChainOfResponsibilityTests {

    private boolean firstWasCalled = false;
    private boolean secondWasCalled = false;
    private boolean thirdWasCalled = false;

    private class First implements ModulusChainLink {

        private final ModulusChainLink next;

        public First(ModulusChainLink next) {

            this.next = next;
        }

        public ModulusResult check(ModulusCheckParams params) {
            firstWasCalled = true;
            return next.check(params);
        }
    }

    private class SecondIsAlwaysResponsible implements ModulusChainLink {

        private final ModulusChainLink next;

        public SecondIsAlwaysResponsible(ModulusChainLink next) {

            this.next = next;
        }

        public ModulusResult check(ModulusCheckParams params) {
            secondWasCalled = true;
            return new ModulusResult(Optional.of(false), Optional.empty());
        }
    }

    private class Third implements ModulusChainLink {

        public ModulusResult check(ModulusCheckParams params) {
            thirdWasCalled = true;
            return new ModulusResult(Optional.of(true), Optional.empty());
        }
    }

    @Test
    public void CanChainTogether() {
        ModulusChainLink chain = new First(new SecondIsAlwaysResponsible(new Third()));

        ModulusResult result = chain.check(new ModulusCheckParams(null, Optional.empty(), Optional.empty(), Optional.empty()));

        assertThat(result.firstCheckResult.get(), is(equalTo(false)));
        assertThat(firstWasCalled, is(equalTo(true)));
        assertThat(secondWasCalled, is(equalTo(true)));
        assertThat(thirdWasCalled, is(equalTo(false)));
    }
}

File: src/test/java/ChainOfResponsibilityTests.kt


import com.github.pauldambra.moduluschecker.ModulusCheckParams
import com.github.pauldambra.moduluschecker.account.BankAccount
import com.github.pauldambra.moduluschecker.chain.ModulusChainLink
import com.github.pauldambra.moduluschecker.chain.ModulusResult
import com.natpryce.hamkrest.assertion.assert
import com.natpryce.hamkrest.equalTo
import org.junit.Test

class ChainOfResponsibilityTests {

    private var firstWasCalled = false
    private var secondWasCalled = false
    private var thirdWasCalled = false

    private inner class First(private val next: ModulusChainLink) : ModulusChainLink {

        override fun check(params: ModulusCheckParams): ModulusResult {
            firstWasCalled = true
            return next.check(params)
        }
    }

    private inner class SecondIsAlwaysResponsible(private val next: ModulusChainLink) : ModulusChainLink {

        override fun check(params: ModulusCheckParams): ModulusResult {
            secondWasCalled = true
            return ModulusResult(false, null)
        }
    }

    private inner class Third : ModulusChainLink {

        override fun check(params: ModulusCheckParams): ModulusResult {
            thirdWasCalled = true
            return ModulusResult(true, null)
        }
    }

    @Test
    fun CanChainTogether() {
        val chain = First(SecondIsAlwaysResponsible(Third()))

        val result = chain.check(ModulusCheckParams(BankAccount("a", "b")))

        assert.that(result.firstCheckResult, equalTo(false))
        assert.that(firstWasCalled, equalTo(true))
        assert.that(secondWasCalled, equalTo(true))
        assert.that(thirdWasCalled, equalTo(false))
    }
}
